### PR TITLE
Create single license scanner for all catalogers

### PR DESCRIPTION
**Pull request by wagoodman**
# Description

Extends work from #3290 by still not leveraging the static license change object (to prevent the GC from holding onto license classifier objects in static space) but also not creating the same scanner often (which results in poor CPU performance). Instead this approach uses a single scanner instance that is scoped to the context object created within the `CreateSBOM()` top-level API function. Since the `licenses.Search()` path is invoked several times, it's important to not create license scanner objects within that function, but instead, pass a new scanner object. This allows for the caller to swap strategies and provide configuration to get different behaviors in license scanning. That being said -- this configuration has NOT been exposed on the API, as the new `licenses.Scanner` is in an internal package, however, in the future we can elevate this to the `pkg/license` package and expose a new builder function on the top-level API `CreateSBOMConfig`, but this has been deferred until later.

- Fixes anchore/grype#2200

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Performance (make Syft run faster or use less memory, without changing visible behavior much)

# Checklist:

- [x] I have added unit tests that cover changed behavior (updated existing tests)
- [x] I have tested my code in common scenarios and confirmed there are no regressions
- [x] I have added comments to my code, particularly in hard-to-understand sections

### CODE REVIEW

**Comment by popey** [comment]
The fix is certainly comparable to the performance of the previous release. 

```
for f in grype-0810 grype-0820 grype-0821 grype-fix; do time ./$f jenkins/jenkins:latest -o json=$f.json; done
 ✔ Vulnerability DB                [no update available]
 ✔ Parsed image sha256:f78b5b90f20c751402aee62cca2342ed20abb9af0d06e45cdc666918ad9f9a00
 ✔ Cataloged contents 62fe8db1f064ce5bfd13cf87d2dcc9083d129b083a43009dde3551393ca9f4fa
   ├── ✔ Packages                        [497 packages]
   ├── ✔ File digests                    [8,512 files]
   ├── ✔ File metadata                   [8,512 locations]
   └── ✔ Executables                     [931 executables]
 ✔ Scanned for vulnerabilities     [156 vulnerability matches]
   ├── by severity: 4 critical, 15 high, 22 medium, 8 low, 92 negligible (15 unknown)
   └── by status:   13 fixed, 143 not-fixed, 0 ignored

real	0m55.700s
user	0m15.853s
sys	0m2.046s
 ✔ Parsed image sha256:f78b5b90f20c751402aee62cca2342ed20abb9af0d06e45cdc666918ad9f9a00
 ✔ Cataloged contents 62fe8db1f064ce5bfd13cf87d2dcc9083d129b083a43009dde3551393ca9f4fa
   ├── ✔ Packages                        [497 packages]
   ├── ✔ File digests                    [8,878 files]
   ├── ✔ File metadata                   [8,878 locations]
   └── ✔ Executables                     [931 executables]
 ✔ Scanned for vulnerabilities     [156 vulnerability matches]
   ├── by severity: 4 critical, 15 high, 22 medium, 8 low, 92 negligible (15 unknown)
   └── by status:   13 fixed, 143 not-fixed, 0 ignored
[0058]  WARN unable to discover java packages from opener: unable to process nested java archive (WEB-INF/lib/javax.inject-1.jar): /usr/share/je
[0067]  WARN unable to parse cpe attributes for elf binary package error=unable to parse Attributes string: failed to parse Attributes="": wfn:

real	1m8.772s
user	0m52.209s
sys	0m2.892s
 ✔ Parsed image sha256:f78b5b90f20c751402aee62cca2342ed20abb9af0d06e45cdc666918ad9f9a00
 ✔ Cataloged contents 62fe8db1f064ce5bfd13cf87d2dcc9083d129b083a43009dde3551393ca9f4fa
   ├── ✔ Packages                        [496 packages]
   ├── ✔ File digests                    [8,878 files]
   ├── ✔ File metadata                   [8,878 locations]
   └── ✔ Executables                     [931 executables]
 ✔ Scanned for vulnerabilities     [156 vulnerability matches]
   ├── by severity: 4 critical, 15 high, 22 medium, 8 low, 92 negligible (15 unknown)
   └── by status:   13 fixed, 143 not-fixed, 0 ignored
[0071]  WARN unable to discover java packages from opener: unable to process nested java archive (WEB-INF/lib/javax.inject-1.jar): /usr/share/je

real	1m28.341s
user	0m54.368s
sys	0m3.399s
 ✔ Parsed image sha256:f78b5b90f20c751402aee62cca2342ed20abb9af0d06e45cdc666918ad9f9a00
 ✔ Cataloged contents 62fe8db1f064ce5bfd13cf87d2dcc9083d129b083a43009dde3551393ca9f4fa
   ├── ✔ Packages                        [496 packages]
   ├── ✔ File digests                    [8,878 files]
   ├── ✔ File metadata                   [8,878 locations]
   └── ✔ Executables                     [931 executables]
 ✔ Scanned for vulnerabilities     [156 vulnerability matches]
   ├── by severity: 4 critical, 15 high, 22 medium, 8 low, 92 negligible (15 unknown)
   └── by status:   13 fixed, 143 not-fixed, 0 ignored
[0053]  WARN unable to discover java packages from opener: unable to process nested java archive (WEB-INF/lib/javax.inject-1.jar): /usr/share/je

real	0m55.420s
user	0m17.839s
sys	0m2.131s
```

File: internal/licenses/context.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package licenses
+
+import (
+	"context"
+)
+
+type licenseScannerKey struct{}
```
Comment (willmurphyscode) [inline internal/licenses/context.go:7]: This is unexported and internal so that callers can't mess with the scanner value in the context? And the getter and setter are exported but internal so that only code that's part of Syft can get or set the context?

Just trying to make sure I understand.

File: internal/licenses/scanner.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	}
+	return &scanner{
+		coverageThreshold: coverageThreshold,
+		scanner:           s.Scan,
+	}
+}
+
+// StaticScanner returns a scanner that uses the built-in license scanner from the licensecheck package.
+// THIS IS ONLY MEANT FOR TEST CODE, NOT PRODUCTION CODE.
+func StaticScanner() Scanner {
```
Comment (willmurphyscode) [inline internal/licenses/scanner.go:40]: Why is this needed for test code? The comment makes it sound super scary and like we should remove it.

File: internal/licenses/scanner.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	}
+
+	content, err := io.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	cov := s.scanner(content)
+	if cov.Percent < s.coverageThreshold {
+		// unknown or no licenses here?
```
Comment (willmurphyscode) [inline internal/licenses/scanner.go:59]: This question reads like a future `TODO`: `TODO: return something that means the scan finished without error but no license could be found?` Is that what we mean here?

**Review by kzantow — APPROVED** [review]

File: internal/licenses/context.go (hunk 1 above)
Comment (wagoodman) [reply internal/licenses/context.go:7]: all correct 👍

File: internal/licenses/scanner.go (hunk 2 above)
Comment (wagoodman) [reply internal/licenses/scanner.go:40]: this is in internal, but the comment is meant to persuade it's use outside of tests (tests are using `context.Background()` and don't have a scanner... and instead of a bunch of tests instantiating license scanners we can use a single one).

File: syft/pkg/cataloger/generic/cataloger.go [hunk 4]
```diff
@@ -154,7 +154,7 @@ func (c *Cataloger) Catalog(ctx context.Context, resolver file.Resolver) ([]pkg.
 	var relationships []artifact.Relationship
 	var errs error
 
-	logger := log.Nested("cataloger", c.upstreamCataloger)
+	lgr := log.Nested("cataloger", c.upstreamCataloger)
```
Comment (willmurphyscode) [inline syft/pkg/cataloger/generic/cataloger.go:157]: I assume this change is unrelated? Keeping a local variable name from shadowing a package name, but nothing to do with fixing the license scanner perf?

File: internal/licenses/scanner.go (hunk 3 above)
Comment (wagoodman) [reply internal/licenses/scanner.go:59]: this is an original comment in the code, the diff doesn't show this since enough changed around it https://github.com/anchore/syft/pull/3348/files/8f58814636f66f48e76f88a07a0171ab937cdc2f#diff-617b280a1fd4309152a54d1b9167be3ff052fd6516650193b1c0612962477348L33

File: syft/pkg/cataloger/generic/cataloger.go (hunk 4 above)
Comment (wagoodman) [reply syft/pkg/cataloger/generic/cataloger.go:157]: unrelated to any performance gains. I was reviewing this code as a part of this work and removed some variable/package shadowing in the process (boy scouting).

File: syft/pkg/cataloger/golang/licenses_test.go [hunk 5]
```diff
@@ -26,6 +28,8 @@ func Test_LocalLicenseSearch(t *testing.T) {
 	loc2 := file.NewLocation("github.com/!cap!o!r!g/!cap!project@v4.111.5/LICENSE.txt")
 	loc3 := file.NewLocation("github.com/someorg/strangelicense@v1.2.3/LiCeNsE.tXt")
 
+	licenseScanner := licenses.StaticScanner()
```
Comment (willmurphyscode) [inline syft/pkg/cataloger/golang/licenses_test.go:31]: This calls a function that loudly documents that it's just for testing. Is there a reason we don't want to make the real license scanner here via `NewDefaultScanner`? Does doing that make the unit tests too slow?

Comment (wagoodman) [reply syft/pkg/cataloger/golang/licenses_test.go:31]: > Does doing that make the unit tests too slow

it's hard to quantify "too slow", but yes, we'd be instantiating a license scanner for every test case this is referenced, which was more than 1 (looks to affect ~33 cases).

Comment (wagoodman) [reply syft/pkg/cataloger/golang/licenses_test.go:31]: disregard, this wasn't reproducible. I'll remove the static scanner

File: internal/licenses/scanner.go (hunk 2 above)
Comment (wagoodman) [reply internal/licenses/scanner.go]: renamed 🎉

**Review by willmurphyscode — APPROVED** [review]
